src/crawl.js

```diff
@@ -0,0 +1,107 @@
+/*
+eslint no-await-in-loop: "off",
+no-constant-condition: ["error", { "checkLoops": false }]
+*/
+
+import jsonld from 'jsonld';
+import assert from 'assert';
+import deepEqual from 'deep-equal';
+import { deepClone, assertValidNode } from './utils/common';
+import { queryNextLookup, dereferenceFromIPFS, parseRDFDocument } from './utils/rdf';
+import { inferh } from './utils/cd';
+import { canon } from './utils/canonicalize';
+import { Namer, fromJsonldjsCg } from './utils/claimgraph';
+
+// Crawl the rdf dataset composed of DIDs and turtle documents on ipfs. Return the graph
+// representing all knowlege obtained while crawling.
+export async function crawl(
+  initialFacts,
+  rules,
+  curiosityQuery,
+  resolveGraph,
+) {
+  // namer is used to ensure blank node hygene whenever adding new claimgraphs
+  const namer = new Namer();
+
+  let facts = deepClone(initialFacts);
+  namer.reallocateNames(facts);
+
+  const lookedup = new Set();
+  const marknew = (term) => {
+    // Add term to lookedup.
+    // Return whether the term was already present in the set.
+    // like https://doc.rust-lang.org/std/collections/btree_set/struct.BTreeSet.html#method.insert
+    const str = canon(term);
+    const isnew = !lookedup.has(str);
+    lookedup.add(str);
+    return isnew;
+  };
+
+  while (true) {
```

Comment:
i think this could be refactored to use promises and recursion, then you dont need to disable await in infinite loop for this

Reply:
Care to go into more detail?

Reply:
For context, I did already put some effort into de-looping (note use of Promise.all). The barrier I ran into is that this process is inherently sequential. The decision of which iris to lookup next depends on all previous results.

Reply:
i think we can overlook that general rule and keep as is for now, thinking about how to deloop it - its doable but it may just add un-necessary complication.